core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -484,13 +485,42 @@ class TransactionStateManager(brokerId: Int,
       responseCallback(responseError)
     }
 
-    replicaManager.appendRecords(
-      newMetadata.txnTimeoutMs.toLong,
-      TransactionLog.EnforcedRequiredAcks,
-      internalTopicsAllowed = true,
-      isFromClient = false,
-      recordsPerPartition,
-      updateCacheCallback)
+    // note that we need to grab the read lock and double check the metadata and coordinator epoch again
+    // until we have appended to log to avoid the scenario that an emigration and immigration happens after
+    // the check when handling the request
+    inReadLock(stateLock) {
```

Comment:
We do. The scenario trying to avoid here is that 1) get the metadata with read lock, 2) this thread gets suspended on the object or the cache map itself gets modified, 3) append to local log which is incorrect.